### ISSUE TITLE
improve speed by factoring sourcevar and matching only levels

### ISSUE THREAD

**Issue by cjyetman** [vincentarelbundock/countrycode#76]
Any interest in verifying/accepting a PR that improves the matching code by factoring the sourcevar first, and then matching only on the levels? It could improve the speed significantly, particularly when sourcevar is a vector of country names with many duplicates (for instance, in a country/year long dataset). Of course, one could pass just the levels and then apply the matches to the long vector, but it would be nice if countrycode handled anything you throw at it in the most efficient way possible. I propose something like the following (which I could wrap up in a PR)...
```
dict <- na.omit(dictionary[, c('regex', destination)])
sourcefctr <- factor(sourcevar)
destination_list <- list()
matches <-
    sapply(levels(sourcefctr), function(x) {
        allmatches <- dict[sapply(dict$regex, function(y) grepl(y, x, perl = TRUE, ignore.case = TRUE)), destination]
        if (length(allmatches) < 1) allmatches <- `class<-`(NA, class(dict[, destination]))
        if (length(allmatches) > 1) destination_list <<- c(destination_list, list(c(x, allmatches)))
        allmatches[length(allmatches)]
    })
destination_vector <- unname(matches[as.numeric(sourcefctr)])
```

**Comment by vincentarelbundock** [comment]
Yes, seems like this could be a significant improvement. A few points/questions:

- I'm quite averse to long one-liners
- Is the output class predictable? How does it relate to inputs?
- How do duplicate matches get resolved?
- Should I be worried about side effects when using ``<<-``?
- Would need to pass all current tests, obviously

**Comment by cjyetman** [comment]
1. one liners can be broken up, sure
2. since the output can only come from one column of countrycode_data, then I assume it's safe to use the class of that column as the class of the NA, i.e. if the output vector must only contain either NA or something from countrycode_data$iso3c, then it's safe to use an NA_character_ because that is the class of countrycode_data$iso3c (is that what you're referring to?)
3. in the above code, I made it create a list (`destination_list`) exactly like the original code does for duplicate matches, that way it doesn't break the warn/error code downstream... likewise, it chooses the **last** match in the `dict$regex` vector, just as the original code (based on my interpretation of how that for loop must run) for consistency... I don't necessarily advocate for those behaviors, but I did it to cause as little trouble as possible
4. I don't know of any specific negative side effects of `<<-`. As far as I understand, it simply searches up through parent environments for the variable it is searching for. I suppose if the variable it was looking for did not exist in the parent environment, then it could continue searching up into higher environments and possibly overwrite a variable with the same name in a parent-parent environment, which could be bad.
5. of course ;-)

**Comment by vincentarelbundock** [comment]
Sounds good.

2. What I meant is if the user supplies a factor, we output a factor. If she supplies a character vector, we return a character vector?
3. Right. The original behavior is far from optimal, but should probably stay as-is until we find a better solution.
4. I see. If there's uncertainty about behavior, can we act conservatively and find an alternative?

Overall, this seems like a great idea; I'm glad this is of interest to you. Thanks!

**Comment by cjyetman** [comment]
2. Both in the current pkg and the above code, the return vector is always the same class as the `dict$destination`, which is typically chr or int. It is not determined by the input `sourcevar`. The factor variable is only used internally by the function, but not passed on to the user. (If the output vector was **really** long, it might make more sense to pass back a factor to the user, but I think that should probably be a user option, not something that would be imposed on them)
3. agreed
4. I will think about it, and investigate options.

**Comment by vincentarelbundock** [comment]
2. yeah. I think I was discussing some nebulous/ideal version of countrycode that exists only in my head. That's fine.

**Comment by cjyetman** [comment]
this is issue has been resolved by merged PR #77. closing issue
